README.rst
.. BANNERSTART
.. Since PyPI does not support raw directives, we remove them from the README
..
.. raw directives are only used to make README fancier on GitHub and do not
.. contain relevant information to be displayed in PyPI, as they are not tied
.. to the current version, but to the current development status
.. raw:: html

    <p align="center">
      <a href="https://github.com/Tesorio/django-anon">
        <img src="https://raw.githubusercontent.com/Tesorio/django-anon/master/icon.svg?sanitize=true" width="128">
      </a>
    </p>

    <h1 align="center">django-anon</h1>
    <p align="center">
      <strong>
        <img src="https://github.githubassets.com/images/icons/emoji/shipit.png" width="16"> Anonymize production data so it can be safely used in not-so-safe environments
      </strong>
    </p>

    <p align="center">
      <a href="https://github.com/Tesorio/django-anon/actions?query=branch%3Amaster">
        <img src="https://github.com/Tesorio/django-anon/workflows/CI/badge.svg?branch=master">
      </a>
      <a href="https://django-anon.readthedocs.io/en/latest/">
        <img src="https://readthedocs.org/projects/pip/badge/?version=latest&style=flat">
      </a>
      <a href="https://github.com/Tesorio/django-anon/blob/master/LICENSE.txt">
        <img src="https://img.shields.io/badge/license-MIT-blue.svg">
      </a>
      <a href="https://pypi.org/project/django-anon/">
        <img src="https://img.shields.io/pypi/v/django-anon.svg?color=blue">
      </a>
    </p>
    
    <p align="center">
      <a href="#installation">
        Install
      </a>
      |
      <a href="https://django-anon.readthedocs.io/en/latest/introduction.html#table-of-contents">
        Read Documentation
      </a>
      |
      <a href="https://pypi.org/project/django-anon/">
        PyPI
      </a>
      |
      <a href="https://github.com/Tesorio/django-anon/blob/master/CONTRIBUTING.rst">
        Contribute
      </a>
    </p>
.. BANNEREND

**django-anon** will help you anonymize your production database so it can be
shared among developers, helping to reproduce bugs and make performance improvements
in a production-like environment.

.. image:: https://raw.githubusercontent.com/Tesorio/django-anon/master/django-anon-recording.gif

.. start-features

Features
========

.. start-features-table

.. csv-table::

   "🚀", "**Really fast** data anonymization and database operations using bulk updates to operate over huge tables"
   "🍰", "**Flexible** to use your own anonymization functions or external libraries like `Faker <https://faker.readthedocs.io/en/latest/index.html>`_"
   "🐩", "**Elegant** solution following consolidated patterns from projects like `Django <https://djangoproject.com/>`_ and `Factory Boy <https://factoryboy.readthedocs.io/en/latest/index.html>`_"
   "🔨", "**Powerful**. It can be used on any projects, not only Django, not only Python. Really!"

.. end-features-table
.. end-features
.. start-table-of-contents

Table of Contents
=================
.. contents::
   :local:

.. end-table-of-contents
.. start-introduction


Installation
------------

.. code::

   pip install django-anon

   
Supported versions
------------------

* Python (2.7, 3.7)
* Django (1.11, 2.2, 3.0)


License
-------

`MIT <https://github.com/Tesorio/django-anon/blob/master/LICENSE>`_

.. end-introduction
.. start-usage


Usage
-----

Use ``anon.BaseAnonymizer`` to define your anonymizer classes:

.. code-block:: python

   import anon

   from your_app.models import Person

   class PersonAnonymizer(anon.BaseAnonymizer):
      email = anon.fake_email
      
      # You can use static values instead of callables
      is_admin = False

      class Meta:
         model = Person

   # run anonymizer: be cautious, this will affect your current database!
   PersonAnonymizer().run()


Built-in functions
~~~~~~~~~~~~~~~~~~

.. code:: python

   import anon

   anon.fake_word(min_size=_min_word_size, max_size=20)
   anon.fake_text(max_size=255, max_diff_allowed=5, separator=' ')
   anon.fake_small_text(max_size=50)
   anon.fake_name(max_size=15)
   anon.fake_username(max_size=10, separator='')
   anon.fake_email(max_size=40, suffix='@example.com')
   anon.fake_url(max_size=50, scheme='http://', suffix='.com')
   anon.fake_phone_number(format='999-999-9999')


Lazy attributes
~~~~~~~~~~~~~~~

Lazy attributes can be defined as inline lambdas or methods, as shown below,
using the ``anon.lazy_attribute`` function/decorator.

.. code-block:: python

   import anon

   from your_app.models import Person

   class PersonAnonymizer(anon.BaseAnonymizer):
      name = anon.lazy_attribute(lambda o: 'x' * len(o.name))

      @lazy_attribute
      def date_of_birth(self):
         # keep year and month
         return self.date_of_birth.replace(day=1)

      class Meta:
         model = Person


The clean method
~~~~~~~~~~~~~~~~

.. code-block:: python

   import anon

   class UserAnonymizer(anon.BaseAnonymizer):
      class Meta:
         model = User

      def clean(self, obj):
         obj.set_password('test')
         obj.save()


Defining a custom QuerySet
~~~~~~~~~~~~~~~~~~~~~~~~~~

A custom QuerySet can be used to select the rows that should be anonymized:

.. code-block:: python

   import anon

   from your_app.models import Person

   class PersonAnonymizer(anon.BaseAnonymizer):
      email = anon.fake_email

      class Meta:
         model = Person

      def get_queryset(self):
         # keep admins unmodified
         return Person.objects.exclude(is_admin=True)


High-quality fake data
~~~~~~~~~~~~~~~~~~~~~~

In order to be really fast, **django-anon** uses it's own algorithm to generate fake data. It is
really fast, but the generated data is not pretty. If you need something prettier in terms of data,
we suggest using `Faker <https://faker.readthedocs.io/en/latest/index.html>`_, which can be used
out-of-the-box as the below:

.. code-block:: python

   import anon

   from faker import Faker
   from your_app.models import Address

   faker = Faker()

   class PersonAnonymizer(anon.BaseAnonymizer):
      postalcode = faker.postalcode

      class Meta:
         model = Address

.. end-usage

Changelog
---------

Check out `CHANGELOG.rst <https://github.com/Tesorio/django-anon/blob/master/CHANGELOG.rst>`_ for release notes

Contributing
------------

Check out `CONTRIBUTING.rst <https://github.com/Tesorio/django-anon/blob/master/CONTRIBUTING.rst>`_ for information about getting involved

----

`Icon <icon.svg>`_ made by `Eucalyp <https://www.flaticon.com/authors/eucalyp>`_ from `www.flaticon.com <https://www.flaticon.com/>`_


CHANGELOG.rst
Changelog
=========

**django-anon**'s release numbering works as follows:

* Versions are numbered in the form **A.B** or **A.B.C.**
* **A.B** is the feature release version number. Each version will be mostly backwards compatible with the previous release. Exceptions to this rule will be listed in the release notes.
* **C** is the patch release version number, which is incremented for bugfix and security releases. These releases will be 100% backwards-compatible with the previous patch release.


Releases
--------

.. contents::
   :local:


master
~~~~~~

* Dropped support for Django 1.8
* ...


0.3.2
~~~~~

* Fixed an infinite loop condition in ``fake_username`` when using the default empty separator


0.3.1
~~~~~

* Fixed bug that happens with newer versions of Django (> 2.2) #63


0.3
~~~

* Updated bulk_update method to use Django's built-in method if available
* Changed default ``max_size`` for ``fake_email`` to ``40``
* Fixed error in ``fake_text`` when ``max_size`` is too short


0.2
~~~

* Added test for Django 3 using Python 3.7 in tox.ini
* Improved performance of fake_text
* Improved performance of BaseAnonymizer.patch_object
* Fix bug with get_queryset not being treated as reserved name
* Improved performance of fake_username
* Removed rand_range argument from fake_username (backwards incompatible)
* Changed select_chunk_size and update_batch_size to saner defaults


0.1
~~~

* Initial release


tests/compat.py
try:
    # stdlib
    from unittest import mock  # noqa: F401
except ImportError:
    # deps
    import mock  # noqa: F401


tests/models.py
# deps
from django.db import models


class PersonAnotherQuerySet(models.QuerySet):
    pass


class Person(models.Model):
    first_name = models.CharField(max_length=255)
    last_name = models.CharField(max_length=255)

    line1 = models.CharField(max_length=255)
    line2 = models.CharField(max_length=255)
    line3 = models.CharField(max_length=255)

    raw_data = models.TextField()

    another_manager = models.Manager.from_queryset(PersonAnotherQuerySet)


def person_factory(**kwargs):
    kwargs.setdefault("first_name", "A")
    kwargs.setdefault("last_name", "B")
    kwargs.setdefault("line1", "X")
    kwargs.setdefault("line2", "Y")
    kwargs.setdefault("line3", "Z")
    kwargs.setdefault("raw_data", '{"access_token": "XYZ"}')

    return Person.objects.create(**kwargs)


tests/settings.py
INSTALLED_APPS = ["tests"]

SECRET_KEY = "j^owl8=_)2do0don9sk@5k7obl!vxm!_404wf%yvk9rp3@a83#"

DATABASES = {"default": {"ENGINE": "django.db.backends.sqlite3", "NAME": ":memory:"}}


anon/base.py
# stdlib
from collections import OrderedDict
from logging import getLogger

# deps
from chunkator import chunkator_page

# local
from anon.compat import bulk_update


logger = getLogger(__name__)


class OrderedDeclaration(object):
    """ Any classes inheriting from this will have an unique global counter
        associated with it. This counter is used to determine the order in
        which fields were declarated

        Idea taken from: https://stackoverflow.com/a/4460034/639465
        Also inspired by https://github.com/FactoryBoy/factory_boy
    """

    global_counter = 0

    def __init__(self):
        self._order = self.__class__.global_counter
        self.__class__.global_counter += 1


class LazyAttribute(OrderedDeclaration):
    def __init__(self, lazy_fn):
        super(LazyAttribute, self).__init__()
        self.lazy_fn = lazy_fn

    def __call__(self, *args, **kwargs):
        return self.lazy_fn(*args, **kwargs)


def lazy_attribute(lazy_fn):
    """ Returns LazyAttribute objects, that basically marks functions that
        should take `obj` as first parameter. This is useful when you need
        to take in consideration other values of `obj`

        Example:

        >>> full_name = lazy_attribute(o: o.first_name + o.last_name)

    """
    return LazyAttribute(lazy_fn)


class BaseAnonymizer(object):
    def run(self, select_chunk_size=None, **bulk_update_kwargs):
        self._declarations = self.get_declarations()

        queryset = self.get_queryset()
        update_fields = list(self._declarations.keys())
        update_batch_size = bulk_update_kwargs.pop(
            "batch_size", self._meta.update_batch_size
        )

        if select_chunk_size is None:
            select_chunk_size = self._meta.select_chunk_size

        if update_batch_size > select_chunk_size:
            raise ValueError(
                "update_batch_size ({}) should not be higher than "
                "select_chunk_size ({})".format(update_batch_size, select_chunk_size)
            )

        # info used in log messages
        model_name = self._meta.model.__name__
        current_batch = 0

        for page in chunkator_page(queryset, chunk_size=select_chunk_size):
            logger.info(
                "Updating {}... {}-{}".format(
                    model_name, current_batch, current_batch + select_chunk_size
                )
            )
            current_batch += select_chunk_size

            objs = []
            for obj in page:
                self.patch_object(obj)
                objs.append(obj)

            bulk_update(
                objs,
                update_fields,
                self.get_manager(),
                **dict(batch_size=update_batch_size, **bulk_update_kwargs)
            )

        if current_batch == 0:
            logger.info("{} has no records".format(model_name))

    def get_meta(self):
        meta = self.Meta()
        if not hasattr(meta, "select_chunk_size"):
            # Chunk size to iterate over
            meta.select_chunk_size = 5000
        if not hasattr(meta, "update_batch_size"):
            # Batch size for bulk updates
            meta.update_batch_size = 200
        return meta

    _meta = property(get_meta)

    def get_manager(self):
        meta = self._meta
        return getattr(meta, "manager", meta.model.objects)

    _manager = property(get_manager)

    def get_queryset(self):
        """ Override this if you want to delimit the objects that should be
            affected by anonymization
        """
        return self.get_manager().all()

    def patch_object(self, obj):
        """ Update object attributes with fake data provided by replacers
        """
        # using obj.__dict__ instead of getattr for performance reasons
        # see https://stackoverflow.com/a/9791053/639465
        fields = [field for field in self._declarations if obj.__dict__[field]]

        for field in fields:
            replacer = self._declarations[field]
            if isinstance(replacer, LazyAttribute):
                # Pass in obj for LazyAttributes
                new_value = replacer(obj)
            elif callable(replacer):
                new_value = replacer()
            else:
                new_value = replacer

            obj.__dict__[field] = new_value

        self.clean(obj)

    def clean(self, obj):
        """ Use this function if you need to update additional data that may
            rely on multiple fields, or if you need to update multiple fields
            at once
        """
        pass

    def get_declarations(self):
        """ Returns ordered declarations. Any non-ordered declarations, for
            example any types that does not inherit from OrderedDeclaration
            will come first, as they are considered "raw" values and should
            not be affected by the order of other non-ordered declarations
        """

        def _sort_declaration(declaration):
            name, value = declaration
            if isinstance(value, OrderedDeclaration):
                return value._order
            else:
                # Any non-ordered declarations come first
                return -1

        declarations = self._get_class_attributes().items()
        sorted_declarations = sorted(declarations, key=_sort_declaration)

        return OrderedDict(sorted_declarations)

    def _get_class_attributes(self):
        """ Return list of class attributes, which also includes methods and
            subclasses, ignoring any magic methods and reserved attributes
        """
        reserved_names = list(BaseAnonymizer.__dict__.keys()) + ["Meta"]

        return {
            name: self.__class__.__dict__[name]
            for name, value in self.__class__.__dict__.items()
            if not name.startswith("__") and name not in reserved_names
        }


anon/compat.py
# deps
from django_bulk_update.helper import bulk_update as ext_bulk_update


def bulk_update(objects, fields, manager, **bulk_update_kwargs):
    """Updates the list of objects using django queryset's
    inbuilt ``.bulk_update()`` method if present, else
    django_bulk_update's ``bulk_update()`` will be used

    :param objects: list of objects that needs to be bulk updated
    :type objects: list[object]
    :param manager: instance of django model manager
    :type manager: models.Manager
    :param bulk_update_kwargs: keyword arguments passed to the ``bulk_update()``
    :return: None
    :rtype: None
    """
    try:
        manager.bulk_update(objects, fields, **bulk_update_kwargs)
    except AttributeError:
        ext_bulk_update(objects, update_fields=fields, **bulk_update_kwargs)


anon/utils.py
# stdlib
import itertools
import random


_WORD_LIST = [
    "a",
    "ab",
    "accusamus",
    "accusantium",
    "ad",
    "adipisci",
    "alias",
    "aliquam",
    "aliquid",
    "amet",
    "animi",
    "aperiam",
    "architecto",
    "asperiores",
    "aspernatur",
    "assumenda",
    "at",
    "atque",
    "aut",
    "autem",
    "beatae",
    "blanditiis",
    "commodi",
    "consectetur",
    "consequatur",
    "consequuntur",
    "corporis",
    "corrupti",
    "culpa",
    "cum",
    "cumque",
    "cupiditate",
    "debitis",
    "delectus",
    "deleniti",
    "deserunt",
    "dicta",
    "dignissimos",
    "distinctio",
    "dolor",
    "dolore",
    "dolorem",
    "doloremque",
    "dolores",
    "doloribus",
    "dolorum",
    "ducimus",
    "ea",
    "eaque",
    "earum",
    "eius",
    "eligendi",
    "enim",
    "eos",
    "error",
    "esse",
    "est",
    "et",
    "eum",
    "eveniet",
    "ex",
    "excepturi",
    "exercitationem",
    "expedita",
    "explicabo",
    "facere",
    "facilis",
    "fuga",
    "fugiat",
    "fugit",
    "harum",
    "hic",
    "id",
    "illo",
    "illum",
    "impedit",
    "in",
    "incidunt",
    "inventore",
    "ipsa",
    "ipsam",
    "ipsum",
    "iste",
    "itaque",
    "iure",
    "iusto",
    "labore",
    "laboriosam",
    "laborum",
    "laudantium",
    "libero",
    "magnam",
    "magni",
    "maiores",
    "maxime",
    "minima",
    "minus",
    "modi",
    "molestiae",
    "molestias",
    "mollitia",
    "nam",
    "natus",
    "necessitatibus",
    "nemo",
    "neque",
    "nesciunt",
    "nihil",
    "nisi",
    "nobis",
    "non",
    "nostrum",
    "nulla",
    "numquam",
    "occaecati",
    "odio",
    "odit",
    "officia",
    "officiis",
    "omnis",
    "optio",
    "pariatur",
    "perferendis",
    "perspiciatis",
    "placeat",
    "porro",
    "possimus",
    "praesentium",
    "provident",
    "quae",
    "quaerat",
    "quam",
    "quas",
    "quasi",
    "qui",
    "quia",
    "quibusdam",
    "quidem",
    "quis",
    "quisquam",
    "quo",
    "quod",
    "quos",
    "ratione",
    "recusandae",
    "reiciendis",
    "rem",
    "repellat",
    "repellendus",
    "reprehenderit",
    "repudiandae",
    "rerum",
    "saepe",
    "sapiente",
    "sed",
    "sequi",
    "similique",
    "sint",
    "sit",
    "soluta",
    "sunt",
    "suscipit",
    "tempora",
    "tempore",
    "temporibus",
    "tenetur",
    "totam",
    "ullam",
    "unde",
    "ut",
    "vel",
    "velit",
    "veniam",
    "veritatis",
    "vero",
    "vitae",
    "voluptas",
    "voluptate",
    "voluptatem",
    "voluptates",
    "voluptatibus",
    "voluptatum",
]


try:
    xrange
except NameError:
    # Python 2/3 proof
    xrange = range


def _cycle_over_sample_range(start, end, sample_size):
    """
    Given a range (start, end), returns a generator that will cycle over a population
    sample with size specified by ``sample_size``
    """
    return itertools.cycle(random.sample(xrange(start, end), sample_size))


def _trim_text(text, separator, max_size):
    limit = min(text.rindex(separator), max_size)
    return text[:limit]


# Holds the maximum size of word sample
_max_word_size = max(len(s) for s in _WORD_LIST)

# Holds a generator that each iteration returns a different word
_word_generator = itertools.cycle(_WORD_LIST)

# Holds the size of smallest word in _WORD_LIST and is used to define bounds
_min_word_size = len(sorted(_WORD_LIST, key=lambda w: len(w))[0])

# Holds a generator that each iteration returns a different number
_number_generator = itertools.cycle("86306894249026785203141")

# Holds a generator for small integers, same as Django's PositiveSmallIntegerField
_small_int_generator = _cycle_over_sample_range(0, 32767, 1000)

# Holds a generator for small signed integers, same as Django's SmallIntegerField
_small_signed_int_generator = _cycle_over_sample_range(-32768, 32767, 1000)

# Holds a generator for integers, same as Django's PositiveIntegerField
_int_generator = _cycle_over_sample_range(0, 2147483647, 10000)

# Holds a generator for signed integers, same as Django's IntegerField
_signed_int_generator = _cycle_over_sample_range(-2147483648, 2147483647, 100000)


def fake_word(min_size=_min_word_size, max_size=20):
    """ Return fake word

    :min_size: Minimum number of chars
    :max_size: Maximum number of chars

    Example:

    >>> import django_anon as anon
    >>> print(anon.fake_word())
    adipisci

    """
    if min_size < _min_word_size:
        raise ValueError("no such word with this size < min_size")

    for word in _word_generator:
        if min_size <= len(word) <= max_size:
            return word


def fake_text(max_size=255, max_diff_allowed=5, separator=" "):
    """ Return fake text

    :max_size: Maximum number of chars
    :max_diff_allowed: Maximum difference (fidelity) allowed, in chars number
    :separator: Word separator

    Example:

    >>> print(anon.fake_text())
    alias aliquam aliquid amet animi aperiam architecto asperiores aspernatur assumenda at atque aut autem beatae blanditiis commodi consectetur consequatur consequuntur corporis corrupti culpa cum cumque cupiditate debitis delectus deleniti deserunt dicta

    """
    if max_diff_allowed < 1:
        raise ValueError("max_diff_allowed must be > 0")

    num_words = max(1, int(max_size / _max_word_size))
    words = itertools.islice(_word_generator, num_words)

    text = separator.join(words)
    try:
        if len(text) > max_size:
            text = _trim_text(text, separator, max_size)
    except ValueError:
        text = text[:max_size]

    return text


def fake_small_text(max_size=50):
    """ Preset for fake_text.

    :max_size: Maximum number of chars

    Example:

    >>> print(anon.fake_small_text())
    Distinctio Dolor Dolore Dolorem Doloremque Dolores

    """
    return fake_text(max_size=max_size).title()


def fake_name(max_size=15):
    """ Preset for fake_text. Also returns capitalized words.

    :max_size: Maximum number of chars

    Example:

    >>> print(anon.fake_name())
    Doloribus Ea

    """
    return fake_text(max_size=max_size).title()


def fake_username(max_size=10, separator=""):
    """ Returns fake username

    :max_size: Maximum number of chars
    :separator: Word separator
    :rand_range: Range to use when generating random number

    Example:

    >>> print(anon.fake_username())
    eius54455

    """
    random_number = str(next(_small_int_generator))
    min_size_allowed = _min_word_size + len(random_number)

    if max_size < min_size_allowed:
        raise ValueError("username must be >= {}".format(min_size_allowed))
    else:
        max_size -= len(random_number)

    return fake_text(max_size, separator=separator) + random_number


def fake_email(max_size=40, suffix="@example.com"):
    """ Returns fake email address

    :max_size: Maximum number of chars
    :suffix: Suffix to add to email addresses (including @)

    Example:

    >>> print(anon.fake_email())
    enim120238@example.com

    """
    min_size_allowed = _min_word_size + len(suffix)

    if max_size + len(suffix) > 254:
        # an email address must not exceed 254 chars
        raise ValueError("email address must not exceed 254 chars")
    elif max_size < min_size_allowed:
        raise ValueError("max_size must be >= {}".format(min_size_allowed))
    else:
        max_size -= len(suffix)

    return fake_username(max_size, separator=".") + suffix


def fake_url(max_size=50, scheme="http://", suffix=".com"):
    """ Returns fake URL

    :max_size: Maximum number of chars
    :scheme: URL scheme (http://)
    :suffix: Suffix to add to domain (including dot)

    Example:

    >>> print(anon.fake_url())
    http://facilis.fuga.fugiat.fugit.harum.hic.id.com

    """
    min_size_allowed = _min_word_size + len(scheme) + len(suffix)

    if max_size < min_size_allowed:
        raise ValueError("max_size must be >= {}".format(min_size_allowed))
    else:
        max_size -= len(scheme) + len(suffix)

    domain = fake_text(max_size=max_size, separator=".") + suffix
    return scheme + domain


def fake_phone_number(format="999-999-9999"):
    """ Returns a fake phone number in the desired format

    :format: Format of phone number to generate

    Example:

    >>> print(anon.fake_phone_number())
    863-068-9424

    """
    number = []
    for char in format:
        if char == "9":
            n = next(_number_generator)
            if not number:
                # do not start phone numbers with zero
                while n == "0":
                    n = next(_number_generator)
            number.append(n)
        else:
            number.append(char)
    return "".join(number)
